### PR TITLE
BUG: optimize: check for NaN optimum in BFGS

### LINKED ISSUE #4793: scipy.optimize.minimize returns success on function that returns NaN
def f(x):                                                                        
        return float('nan')

    res = scipy.optimize.minimize(f, 1)
    print(res)

will show ```Optimization terminated successfully.```  and ```success: True```, while obviously the minimum has not be found.

Explicitly setting the method, to, e.g., ```method=Nelder-mead``` does return ```success: False```, which *is* what one would expect.

### PR DESCRIPTION
**Pull request by larsmans**
New take on #5125 to fix #4793, at least for BFGS. Ping @pv, @josef-pkt. (GitHub wouldn't let me reopen the old PR after a force-push.) The solution turns out to be very simple.

### CODE REVIEW

File: scipy/optimize/tests/test_optimize.py [hunk 1]
```diff
@@ -384,6 +384,20 @@ def test_bfgs_nan(self):
             x = optimize.fmin_bfgs(func, x0, fprime, disp=False)
             assert_(np.isnan(func(x)))
 
+    def test_bfgs_nan_return(self):
+        # Test corner cases where fun returns NaN. See gh-4793.
+
+        # First case: NaN from first call.
+        func = lambda x: np.nan
+        result = optimize.minimize(func, 0)
+        assert_(np.isnan(result['fun']) and result['success'] is False)
```
Comment (josef-pkt) [inline scipy/optimize/tests/test_optimize.py]: I would have written two asserts, so it's more visible which one fails. (second could be assert_equal

**Comment by josef-pkt** [comment]
looks good to me for bfgs. 
However, I thought this could also go into the minimize top function itself it improves consistency across optimizers. (to make sure all have success == False)

**Comment by larsmans** [comment]
@josef-pkt Fixing it in `minimize` would leave `fmin_bfgs`, `fmin_l_bfgs_b` and `fmin_cg` without the check.

File: scipy/optimize/tests/test_optimize.py (hunk 1 above)
Comment (larsmans) [inline scipy/optimize/tests/test_optimize.py]: Done.

**Comment by larsmans** [comment]
Merged and made a new issue for the other optimizers.

**Pull request merged**